File: src/main/kotlin/org/kite9/diagram/common/elements/mapping/AbstractBaseCornerVertices.java
package org.kite9.diagram.common.elements.mapping;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.kite9.diagram.common.elements.vertex.MultiCornerVertex;
import org.kite9.diagram.common.fraction.BigFraction;
import org.kite9.diagram.common.objects.OPair;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D;

public abstract class AbstractBaseCornerVertices extends AbstractCornerVertices {

	public static final OPair<BigFraction> FULL_RANGE = new OPair<>(BigFraction.Companion.getZERO(), BigFraction.Companion.getONE());
	
	private final Map<OPair<BigFraction>, MultiCornerVertex> elements = new HashMap<>();
	
	public AbstractBaseCornerVertices(DiagramElement rootContainer, OPair<BigFraction> cx, OPair<BigFraction> cy, int depth) {
		super(rootContainer, cx, cy, depth);
	}


	protected MultiCornerVertex getExistingVertex(OPair<BigFraction> d) {
		return elements.get(d);
	}
	
	/**
	 * There are no duplicates for independent container vertices.
	 */
	@Override
	public MultiCornerVertex mergeDuplicates(MultiCornerVertex cv, RoutableHandler2D rh) {
		return cv;
	}

	@Override
	public MultiCornerVertex createVertex(BigFraction x, BigFraction y) {
		return createVertexHere(x, y, elements);
	}

	@Override
	protected AbstractCornerVertices getTopContainerVertices() {
		return this;
	}

	@Override
	public Collection<MultiCornerVertex> getAllAscendentVertices() {
		return elements.values();
	}

	public Collection<MultiCornerVertex> getVerticesAtThisLevel() {
		return elements.values();
	}
	
	
	@Override
	public Collection<MultiCornerVertex> getAllDescendentVertices() {
		Collection<MultiCornerVertex> out = super.getAllDescendentVertices();
		out.addAll(elements.values());
		return out;
	}
}

File: src/main/kotlin/org/kite9/diagram/common/elements/mapping/AbstractBaseCornerVertices.kt
package org.kite9.diagram.common.elements.mapping

import org.kite9.diagram.common.elements.vertex.MultiCornerVertex
import org.kite9.diagram.common.fraction.LongFraction
import org.kite9.diagram.common.fraction.LongFraction.Companion.ONE
import org.kite9.diagram.common.fraction.LongFraction.Companion.ZERO
import org.kite9.diagram.common.objects.OPair
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D

abstract class AbstractBaseCornerVertices(
    rootContainer: DiagramElement,
    cx: OPair<LongFraction>,
    cy: OPair<LongFraction>,
    depth: Int
) : AbstractCornerVertices(
    rootContainer, cx, cy, depth
) {
    private val elements: MutableMap<OPair<LongFraction>, MultiCornerVertex> = HashMap()

    /**
     * There are no duplicates for independent container vertices.
     */
    override fun mergeDuplicates(cv: MultiCornerVertex, rh: RoutableHandler2D): MultiCornerVertex {
        return cv
    }

    override fun createVertex(x: LongFraction, y: LongFraction): MultiCornerVertex {
        return createVertexHere(x, y, elements)
    }

    override fun getTopContainerVertices(): AbstractCornerVertices {
        return this
    }

    override fun getAllAscendentVertices(): MutableCollection<MultiCornerVertex> {
        return elements.values.toMutableList()
    }

    override fun getVerticesAtThisLevel(): Collection<MultiCornerVertex> {
        return elements.values
    }

    override fun getAllDescendentVertices(): MutableCollection<MultiCornerVertex> {
        val out = super.getAllDescendentVertices()
        out.addAll(elements.values)
        return out
    }

    companion object {
		val FULL_RANGE = OPair(ZERO, ONE)
    }
}